components/raftstore/src/store/peer.rs

```diff
@@ -3780,7 +3780,13 @@ where
         ctx.raft_metrics.propose.transfer_leader += 1;
 
         let transfer_leader = get_transfer_leader_cmd(&req).unwrap();
-        let peer = transfer_leader.get_peer();
+        let prs = self.raft_group.raft.prs();
+
+        let peer = transfer_leader
+            .get_peers()
+            .iter()
+            .max_by_key(|c| prs.get(c.id).map_or(0, |pr| pr.matched))
```

Comment:
If the preferred node is the first one and all their matcheds are the same, then the first one should be chose. `max_by_key` seems to use the last one.

Reply:
When using the `peers` field, we preferred a random pick node if all `pr.matched` are the same.

Reply:
Always using the last one is not random apparently. As I recall, PD will put the preferred one to the first of the list.

Reply:
> Always using the last one is not random apparently

The position in the peer list is fixed but the node id may be random

> PD will put the preferred one to the first of the list

Using the first one is okay to me if there are such API promise

Reply:
> As I recall, PD will put the preferred one to the first of the list.

I'm not sure where you are referring to. In [pd#4401](https://github.com/tikv/pd/pull/4401/files) I just return and copy the full candidates list, without sorting or shuffling. So maybe a random pick is better here?

Reply:
Maybe check `transfer_leader` first? If `transfer_leader` doesn't have enough progress, then a random pick.